### ISSUE TITLE
SIGSEGV within com.badlogic.graphics.gdx.Mesh (workaround described in reply below)

### ISSUE THREAD

**Issue by robertlight** [libgdx/libgdx#5610]
**Issue details**
I have the following:

```
LwjglAWTCanvas gdxCanvas = new LwjglAWTCanvas( myApp ) ;
JPanel mySwingJPanel = new JPanel();
mySwingJPanel.add( gdxCanvas.getCanvas(), BorderLayout.CENTER); 
```
later on I hide my canvas:
`gdxCanvas.getCanvas().setVisible(false);`

This works fine.

When I set my canvas visible again:
`gdxCanvas.getCanvas().setVisible(true);`

I get a SIGSEGV and bail out of the JVM

The SIGSEGV happens in the class com.badlogic.graphics.gdx.Mesh, Line: 292
`Gdx.gl20.glDrawElements(primitiveType, count, 5123, offset * 2);`

at the time of the crash, primitiveType==1, count==8, offset==0

NOTE:  If I put a breakpoint at this line and just resume everytime I get to this line, I do NOT get the SIGSEGV - indicating that there might be a timing issue.   Then again... I tried to put a Thread.sleep(20) before the call to Gdx.gl20.glDrawElements....and I couldn't get it avoid the SIGSEGV..

Can anyone suggest a fix and/or a workaround?

Version of LibGDX and/or relevant dependencies:
```
gdx-1.9.9.jar
jinput-platform-2.0.5-natives-windows.jar
gdx-backend-headless-1.9.9.jar
gdx-tools-1.9.9.jar
jlayer-1.0.1-gdx.jar
gdx-backend-lwjgl-1.9.9.jar
jinput-2.0.5.jar
jorbis-0.0.17.jar
gdx-freetype-1.9.9.jar
jinput-platform-2.0.5-natives-linux.jar
jutils-1.0.0.jar
gdx-platform-1.9.9-natives-desktop.jar
jinput-platform-2.0.5-natives-osx.jar
lwjgl-2.9.2.jar
lwjgl-platform-2.9.2-natives-osx.jar
lwjgl_util-2.9.2.jar
lwjgl-platform-2.9.2-natives-linux.jar
lwjgl-platform-2.9.2-natives-windows.jar
```
Affected platform:  MacOSX

I'm attaching an IntelliJ project that demonstrates the SIGSEGV in a comment below.

The workaround seems to be to make sure that all Model objects are have their .dispose() methods called and then recreated in the create() method.  IMPORTANT:  all models held in an AssetManager also need to be disposed by calling assetMgr.dispose() and then reloading them from disk.

**Comment by intrigus** [comment]
This needs a minimal sample program.

**Comment by robertlight** [comment]
The test code just opens a JFrame, installs a JPanel in the RootContentPane and then puts the lwjgl canvas into the JPanel.

It then starts two timers, one that fires in 5 seconds that removes the lwjgl canvas and then one that fires 5 seconds later that adds the lwjgl canvas back into the JPanel.

There is a commented out line that does a clear of an ArrayList that, if uncommented, the SIGSEGV goes away.

[test.tar.gz](https://github.com/libgdx/libgdx/files/3105531/test.tar.gz)

**Comment by robertlight** [comment]
Here may be a hint... I'm using assets loaded from a .g3db file.   I can't seem to do a model.dispose() on a model retrieved from an AssetManager and somehow if I reuse the Model object, I cause a SIGSEGV.   Should I be reloading my asset during the create() method?

**Comment by Darkyenus** [comment]
libGDX generally does not check for use-after-dispose. And OpenGL drivers shouldn't crash when they are supplied with non-existent (=freed) resource, so this should not be a problem. In reality, they do crash, and there is not much we can do about it without sacrificing performance.

**Comment by robertlight** [comment]
I wasn't using "disposed" models...I was using UNdisposed models...but models made before the create() method was called....that seems to be the source of the SIGSEGV.

I believe this is something that libgdx could at least throw a useful exception rather that wait for a mysterious SIGSEGV.

**Comment by Darkyenus** [comment]
I have looked at the code more closely:
- Manipulating Swing/AWT components in random threads is not allowed, so I have wrapped what your code did in timer tasks in `SwingUtilities.invokeLater`
- For me it crashes *during* panel removal, in `glFinish`
- Uncommenting `models.clear()` does not fix the issue

Adding some delays to the code makes it sometimes work, but it is not reliable. This looks like a lwjgl or a driver bug. I have no idea how to fix this at the libGDX level.

Also, in your initial report you mention `setVisible`, but your code does not call it on canvas. Instead it removes the canvas from AWT tree. Changing that to `setVisible` calls does not lead to any crashes, but the visibility seems to be ignored (which might be a bug). A workaround that did work properly was to set the size of the canvas to zero.

Mixing AWT with OpenGL was never very reliable, so unless you really need it because of some specific AWT/Swing components (or for accessibility), I suggest staying in OpenGL with scene2d.

Btw. [`Mesh:292`](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/graphics/Mesh.java#L292) does not contain `glDrawElements`?

**Comment by robertlight** [comment]
My mistake... I was operating off of the decompiler's code, not the actual code.
The SIGSEGV was caused on [Mesh:512](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/graphics/Mesh.java#L512)

I think I found that the setVisiblity() was not effective in blanking out the frame so I went to the technique of deleting the JFrame entirely and recreating it.  That's when I discovered that if you had Model objects laying around from a previous libGDx display, that you got the SIGSEGV.

My libGDx world lives within a much larger Swing application - so that's why I need it to "go away" every once in a while and restart up when needed.  I'm not doing things in separate threads like I hacked into the test code above.  I'll try to set it up without the Timers and just make it more like my larger application with the libGDx frame in one tabbed panel and then have something else in another of the CardLayout panels.
